### DISCUSSION TITLE
make api params available without explicit h3 import

### DISCUSSION THREAD

**Discussion by philippdormann** [nuxt/nuxt#18311] [Ideas]
simplify access to query params (body params, path params)

I currently have to add a request parser (h3) as a dependency
```
import { useQuery } from 'h3'
export default (req, res) => {
    const query = useQuery(req)
    console.log(query);
    return query
}
```

this should directly be integrated into nuxt framework

**Comment by philippdormann** [comment]
my current scalable/ clean solution is to add a middleware for query parsing
```
// server/middleware/query.ts
import { useQuery } from 'h3'
export default async (req, res) => {
    const query = useQuery(req)
    req.query = query
}
```

so that I can use query params like
```
// server/api/users.ts
export default (req, res) => {
    return req.query
}
```

**Reply by justindasilva** [reply]
When I added the `middleware/query.ts` file, I get this error:
```
handle is not a function
  at file://./node_modules/h3/dist/index.mjs:111:24 
```

Any ideas how to fix?

**Comment by pi0** [comment]
Hi @philippdormann. Sure you can opt-in to use a middleware parsing query on all routes (while not recommended) but this design is intentional to avoid adding extra parsing logic when most of the requests hit SSR middleware and using vue-router for parsing url. h3 is already a dependency of nitro and has no footprint to your project by using. It will even be tree-shaked finally :)

**Reply by justindasilva** [reply]
Thanks for the advice. It's true that not all of my routes will need to make use of `useQuery`, even though most will. But I will get into habit of including it only when necessary.

This is working for me as expected now:
```
import { useQuery } from 'h3'
import { MyService } from '#services'

export default async (req) => {

  if (req.method !== 'GET') {
    throw new Error('This should be a GET request')
  }

  const query = useQuery(req)
  const data = await MyService.getAll(query.view)

  return data
}
```

I wonder if something better can be done to distinguish GET and POST requests? I'm using example I see here: https://github.com/nuxt/framework/discussions/1102

**Reply by philippdormann** [reply]
hm, so how are we supposed to access body and query params then?
I still need to import h3's useQuery/ useBody in my API endpoints (handlers), right?

(my question was just about the server/ backend logic, not vue-router)
I hoped Nuxt Framework has a similar route to Next.js - where these query parameters and request body parsing does work 🤔

highly interested on your decision to keep Nuxt this lightweight/ bare bones ;)
is it purely performance-reasons (avoiding the parsing)?

**Reply by pi0** [reply]
@philippdormann Yes mainly performance and scaling. This approach allows you to separate your server logic (like `useDatabase`) into small compostables and also not forced to use `h3` implementation (`req` and `res` are pure Node.js API). BTW I'm pretty sure we can even have a nuxt module that adds this functionality to fill `req` object giving the same experience of `next.js` if you prefer.

**Reply by philippdormann** [reply]
yeah, cool. makes total sense from a scaling point of view 👍
it might be a good idea adding these infos to the docs though
